src/schedule/zephyr_twb_schedule.c

```diff
@@ -0,0 +1,487 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright(c) 2024 Intel Corporation. All rights reserved.
+ *
+ * Author: Adrian Bonislawski <adrian.bonislawski@intel.com>
+ */
+
+#include <rtos/task.h>
+#include <stdint.h>
+#include <sof/lib/uuid.h>
+#include <sof/schedule/schedule.h>
+#include <sof/schedule/ll_schedule.h>
+#include <sof/schedule/twb_schedule.h>
+#include <sof/schedule/ll_schedule_domain.h>
+#include <sof/trace/trace.h>
+#include <rtos/wait.h>
+#include <rtos/interrupt.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/kernel/thread.h>
+
+LOG_MODULE_REGISTER(twb_schedule, CONFIG_SOF_LOG_LEVEL);
+
+/* e93326d8-0d14-4bf0-bcb9-e063d3d80136 */
+DECLARE_SOF_UUID("twb-schedule", twb_sched_uuid, 0xe93326d8, 0x0d14, 0x4bf0,
+		 0xbc, 0xb9, 0xe0, 0x63, 0xd3, 0xd8, 0x01, 0x36);
+
+DECLARE_TR_CTX(twb_tr, SOF_UUID(twb_sched_uuid), LOG_LEVEL_INFO);
+
+struct scheduler_twb_data {
+	struct list_item tasks;		/* list of active TWB tasks */
+	struct task ll_tick_src;	/* LL task - source of TWB tick */
+};
+
+struct task_twb_data {
+	k_tid_t thread_id;		/* zephyr thread ID */
+	k_thread_stack_t __sparse_cache *p_stack; /* pointer to thread stack */
+	int32_t thread_prio;		/* thread default priority */
+	uint32_t cycles_granted;	/* cycles budget for the task */
+	uint32_t cycles_consumed;	/* cycles consumed by the task */
+	uint64_t cycles_ref;		/* reference cycles for the task */
+};
+
+/* Single CPU-wide lock
+ * as each per-core instance if TWB-scheduler has separate structures, it is enough to
+ * use irq_lock instead of cross-core spinlocks
+ */
+static inline unsigned int scheduler_twb_lock(void)
+{
+	return irq_lock();
+}
+
+static inline void scheduler_twb_unlock(unsigned int key)
+{
+	irq_unlock(key);
+}
+
+/**
+ * @brief Callback function for the TWB scheduler task.
+ *
+ * This function is called when the TWB scheduler task runs out of budget.
+ * It lowers the priority of the thread and sets the time slice to 0.
+ *
+ * @param thread Pointer to the thread structure.
+ * @param data   Pointer to additional data (not used).
+ */
+static void scheduler_twb_task_cb(struct k_thread *thread, void *data)
+{
+	tr_dbg(&twb_tr, "TWB task %p out of budget, lowering priority", data);
+
+	k_thread_priority_set(thread, CONFIG_TWB_THREAD_LOW_PRIORITY);
+	k_thread_time_slice_set(thread, 0, &scheduler_twb_task_cb, data);
```

Comment:
For now - yes because this is not described in TWB documentation so Im using default value here (priority is the most important part here) but this is something we can discuss